stdlib/internal/promql/join.go

```diff
@@ -0,0 +1,553 @@
+	rowB := make(map[string]values.Value, 8)
+
+	_, ra, ta := iterA.next()
+	i, rb, tb := iterB.next()
+
+	var firstRow = true
+
+NEXT:
+	if ra == -1 || rb == -1 {
+		goto DONE
+	}
+
+	if ta < tb {
+		_, ra, ta = iterA.next()
+		goto NEXT
+	}
+	if ta > tb {
+		i, rb, tb = iterB.next()
+		goto NEXT
+	}
+
+	iterA.row(rowA)
+
+	for r, t := rb, tb; r != -1 && t == ta; _, r, t = iterB.next() {
+		iterB.row(rowB)
+
+		// Evaluate fn over both input rows
+		obj, err := c.fn.Eval(
+			c.ctx,
+			values.NewObjectWithValues(rowA),
+			values.NewObjectWithValues(rowB),
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		// Build schema if this is the first row being joined
+		if firstRow {
+			schema := make([]flux.ColMeta, 0, obj.Len())
+			obj.Range(func(name string, v values.Value) {
+				schema = append(schema, flux.ColMeta{
+					Label: name,
+					Type:  execute.ConvertFromKind(v.Type().Nature()),
+				})
+			})
+			sort.Slice(schema, func(i, j int) bool {
+				return schema[i].Label < schema[j].Label
+			})
+			for _, col := range schema {
+				if _, err := builder.AddCol(col); err != nil {
+					return nil, err
+				}
+			}
+			firstRow = false
+		}
+
+		// Check fn does not update the group key values.
+		// TODO(josh): This should be caught during planning.
+		// Remove this when the planner is made schema aware.
+		for _, col := range key.Cols() {
+			v, ok := obj.Get(col.Label)
+			if !ok {
+				return nil, errors.New(codes.Invalid, "argument 'fn' may not modify group key")
+			}
+			if !v.Equal(key.LabelValue(col.Label)) {
+				return nil, errors.New(codes.Invalid, "argument 'fn' may not modify group key")
+			}
+		}
+
+		// The record obtained from calling fn may be added to output
+		if err := appendRowToBuilder(builder, obj); err != nil {
+			return nil, err
+		}
+	}
+
+	iterB.reset(i, rb)
+	_, ra, ta = iterA.next()
+	goto NEXT
+DONE:
+	return builder.Table()
+}
+
+func appendRowToBuilder(builder *execute.ColListTableBuilder, obj values.Object) error {
+	var err error
+	obj.Range(func(name string, v values.Value) {
+		idx := execute.ColIdx(name, builder.Cols())
+		if idx < 0 {
+			err = errors.Newf(codes.NotFound, "column %s not found", name)
+			return
+		}
+		if err = builder.AppendValue(idx, v); err != nil {
+			return
+		}
+	})
+	return err
+}
+
+// iterator for a list of column readers
+type colReaderIterator struct {
+	readers []flux.ColReader
+	timeCol int
+	i, r    int
+}
+
+func (iter *colReaderIterator) next() (int, int, int64) {
+	iter.i++
+	reader := iter.readers[iter.r]
+	if iter.i < reader.Len() {
+		return iter.i, iter.r, reader.Times(iter.timeCol).Value(iter.i)
+	}
+	iter.r++
+	if iter.r < len(iter.readers) {
+		iter.i = 0
+		return iter.i, iter.r, iter.readers[iter.r].Times(iter.timeCol).Value(iter.i)
+	}
+	return -1, -1, -1
+}
+
+func (iter *colReaderIterator) reset(i, r int) {
+	iter.i = i
+	iter.r = r
+}
+
+func (iter *colReaderIterator) row(m map[string]values.Value) {
+	for k := range m {
```

Comment:
I don't think this should ever be needed. The tables will always have a consistent schema so you are always going to have the same values that you are filling in.

Reply:
I don't think this is true at the moment. For instance, if `fn` returns two rows of different types (perhaps by an if-then expression), the type system currently cannot catch that. I'll add a comment so that it's clear.

Reply:
Ah I didn't think of that. Never mind then.